src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -496,72 +517,92 @@ func (c controller) postInstallConfigs(ctx context.Context) (error, map[string]i
 	return nil, nil
 }
 
-func (c controller) waitForOLMOperators(ctx context.Context) error {
+func (c *controller) isStageTimedOut(stage models.FinalizingStage) func() bool {
+	return func() bool {
+		cluster, err := c.ic.GetCluster(context.TODO(), false)
+		if err != nil {
+			c.log.WithError(err).Error("failed to get cluster")
+			return false
+		}
+		return cluster != nil && cluster.Progress != nil && cluster.Progress.FinalizingStage == stage && cluster.Progress.FinalizingStageTimedOut
+	}
+}
+
+func (c *controller) waitForOLMOperators(ctx context.Context) error {
 	var operators []models.MonitoredOperator
-	var err error
+	var err, retryErr error
+	var errs []error
 
 	// In case the timeout occur, we have to update the pending OLM operators to failed state,
 	// so the assisted-service can update the cluster state to complete.
 	defer func() {
-		if err == nil {
+		if errs == nil {
 			return
 		}
-		updateFunc := func() bool {
-			if err = c.updatePendingOLMOperators(ctx); err != nil {
-				return KeepWaiting
-			}
-			return ExitWaiting
+		updateFunc := func() error {
+			return c.updatePendingOLMOperators(context.TODO())
```

Comment:
Not sure why not to use same context as in lline 545

Reply:
OK. Will use